### ISSUE TITLE
Blazor controlled binding to integer values does not work

### ISSUE THREAD

**Issue by MarvinKlein1508** [dotnet/aspnetcore#31003]
### Describe the bug
I am trying to add more controlled binding to an `InputSelect `razor component. What I am trying to do is to execute a method when the value of the property has changed. However, my value is of type `int`.

```
<InputSelect Value="Input.AddressId"
             ValueChanged="@((string s) => AddressChangedAsync(s))"
             ValueExpression="@(() => Input.AddressId)"
             class="form-control">
    <option value="0">--- Keine Lieferanschrift ---</option>
    @foreach (var address in _customer_addresses)
    {
        <option value="@address.Id">address.ToString()</option>
    }
</InputSelect>

```

```
private Task AddressChangedAsync(string value)
{
    _ = int.TryParse(value, out int addressId);

    Input.AddressId= addressId;
    Input.Address= _customer_addresses.Where(x => x.Id== addressId).FirstOrDefault();

    return Task.CompletedTask;
}
```

However, the program will not compile and throw the exception mentioned below. As it says I should declare TValue. So I added `TValue="int"` to the `InputSelect`. Now I get the second error. The same error with `TValue="typeof(int)"`

This is really strange because InputSelect should support integer types since .NET 5.

### Exceptions (if any)
CS0411	The type arguments for method 'TypeInference.CreateInputSelect_45<TValue>(RenderTreeBuilder, int, int, TValue, int, EventCallback<TValue>, int, Expression<Func<TValue>>, int, object, int, RenderFragment)' cannot be inferred from the usage. Try specifying the type arguments explicitly.

CS1503: Argument "1": Converting from "int" to "System.ReadOnlySpan<char>" is not possible.

### Further technical details
- ASP.NET Core version 5.0.4
.NET SDK (gemäß "global.json"):
 Version:   5.0.201
 Commit:    a09bd5c86c

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.201\

Host (useful for support):
  Version: 5.0.4
  Commit:  f27d337295

.NET SDKs installed:
  3.1.400-preview-015203 [C:\Program Files\dotnet\sdk]
  3.1.407 [C:\Program Files\dotnet\sdk]
  5.0.100-rc.1.20452.10 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]
  5.0.103 [C:\Program Files\dotnet\sdk]
  5.0.104 [C:\Program Files\dotnet\sdk]
  5.0.200-preview.21079.7 [C:\Program Files\dotnet\sdk]
  5.0.201 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-rc.1.20451.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-rc.1.20451.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.1.20452.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Comment by MarvinKlein1508** [comment]
I just found out that I need to declare my type as `int?` instead of `int`. Not sure why but with this change it works now.

**Comment by SteveSandersonMS** [comment]
> CS1503: Argument "1": Converting from "int" to "System.ReadOnlySpan" is not possible.

I think that error is coming from your own code:

```
_ = int.TryParse(value, out int addressId);
```

You must have changed the argument type from `string` to `int` to trigger this.

If you're still having trouble, could you please post a complete and minimal repro? The code above doesn't compile for other reasons.

**Comment by MarvinKlein1508** [comment]
Hi @SteveSandersonMS 
no this error is not coming from this line. I created a small [repository](https://github.com/MarvinKlein1508/BlazorCustomBinding) which is unable to compile.

It will compile when I enable nullable for the project and define my `AddressId `in `Customer.cs` as `int?`

**Comment by SteveSandersonMS** [comment]
Sorry, I haven't been able to repro a problem. The code in the repro you've linked to seems to compile perfectly well for me, using .NET Core 5.0.104.

I did notice what appears to be a mistake in the code though (which doesn't stop it from compiling, but it does behave strangely at runtime):

```razor
        <InputSelect class="form-control"
                     Value="Customer.CustomerId"
                     ValueExpression="() => Customer.CustomerId"
                     ValueChanged="(int addressId) => AddressChangedAsync(addressId)">
``` 

I think `CustomerId` should be `AddressId`, like this:

```razor
        <InputSelect class="form-control"
                     Value="Customer.AddressId"
                     ValueExpression="() => Customer.AddressId"
                     ValueChanged="(int addressId) => AddressChangedAsync(addressId)">
```

**Comment by MarvinKlein1508** [comment]
@SteveSandersonMS 

strange now it works. I have no clue why. When I uploaded the demo repo yesterday it wasn't working. Hmm seems like a restart of Visual Studio has fixed this.

What I did:
1. Created new Blazor server project
2. Removed WeaterService and Weaterforecast
3. Removed Counter and FetchData.razor
4. Removed these files from navigation
5. Removed weather service from Startup
6. Created my two models
7. Created my sample data in Index.razor
8. Created my EditForm

I tried a different settings now (VS Preview, normal VS) Wiith nullable enabled and without. Seems to work now. But not sure why.

Oh and yes you were right. It must be `AddressId` instead of `CustomerId`. When I experience this issue again, I'll let you know.
